scc/tests/args.rs

```diff
@@ -25,226 +36,166 @@ const TEST_DIRECTORY: TestArg = TestArg {
+    #[values(Some(&TEST_SCRIPT_PATHS), None)] script_paths_file: Option<&TestArg>,
+    #[values(Some("none"), None)] warnings: Option<&str>,
+    #[values(Some(true), None)] optimize: Option<bool>,
+    #[values(Some(4), None)] threads: Option<u8>,
+    #[values(Some(false), None)] no_testonly: Option<bool>,
+    #[values(Some(false), None)] no_breakpoint: Option<bool>,
+    #[values(Some(true), Some(false), None)] profile: Option<bool>,
+) {
 }
 
-mod standard {
+#[cfg(test)]
+mod tests {
     use super::*;
 
     #[test]
-    fn test_empty() {
-        test_order(&[]);
-    }
-    #[test]
-    fn test_fd() {
-        test_order(&[TEST_FILE, TEST_DIRECTORY]);
-    }
-    #[test]
-    fn test_fdf() {
-        test_order(&[TEST_FILE, TEST_FILE, TEST_FILE]);
-    }
-    #[test]
-    fn test_ffd() {
-        test_order(&[TEST_FILE, TEST_FILE, TEST_DIRECTORY]);
-    }
-    #[test]
-    fn test_fdd() {
-        test_order(&[TEST_FILE, TEST_DIRECTORY, TEST_DIRECTORY]);
-    }
-    #[test]
-    fn test_fddf() {
-        test_order(&[TEST_FILE, TEST_DIRECTORY, TEST_DIRECTORY, TEST_FILE]);
-    }
-    #[test]
-    fn test_fdfd() {
-        test_order(&[TEST_FILE, TEST_DIRECTORY, TEST_DIRECTORY, TEST_FILE]);
-    }
-    #[test]
-    fn test_fddfd() {
-        test_order(&[TEST_FILE, TEST_DIRECTORY, TEST_DIRECTORY, TEST_FILE, TEST_DIRECTORY]);
-    }
-    #[test]
-    fn test_df() {
-        test_order(&[TEST_DIRECTORY, TEST_FILE]);
-    }
-    #[test]
-    fn test_dfd() {
-        test_order(&[TEST_DIRECTORY, TEST_FILE, TEST_DIRECTORY]);
-    }
-    #[test]
-    fn test_ddf() {
-        test_order(&[TEST_DIRECTORY, TEST_DIRECTORY, TEST_FILE]);
-    }
-    #[test]
-    fn test_dff() {
-        test_order(&[TEST_DIRECTORY, TEST_FILE, TEST_FILE]);
-    }
-    #[test]
-    fn test_dffd() {
-        test_order(&[TEST_DIRECTORY, TEST_FILE, TEST_FILE, TEST_DIRECTORY]);
-    }
-    #[test]
-    fn test_dfdf() {
-        test_order(&[TEST_DIRECTORY, TEST_FILE, TEST_FILE, TEST_DIRECTORY]);
+    fn get_help() {
+        println!(
+            "{}",
+            Opts::get_parser()
+                .run_inner(bpaf::Args::from(&["--help"]))
+                .unwrap_err()
+                .unwrap_stdout()
+        );
     }
+
     #[test]
-    fn test_dffdf() {
-        test_order(&[TEST_DIRECTORY, TEST_FILE, TEST_FILE, TEST_DIRECTORY, TEST_FILE]);
+    fn check_options() {
+        Opts::get_parser().check_invariants(false)
     }
-}
 
-mod cybercmd {
-    use super::*;
+    #[apply(file_directory_orders)]
+    fn standard(
+        f_args: &[TestArg],
+        cache_file: Option<&TestArg>,
+        cache_dir: Option<&TestArg>,
+        script_paths_file: Option<&TestArg>,
+        warnings: Option<&str>,
+        optimize: Option<bool>,
+        threads: Option<u8>,
+        no_testonly: Option<bool>,
+        no_breakpoint: Option<bool>,
+        profile: Option<bool>,
+    ) {
+        std::panic::set_hook(Box::new(|a| println!("{:#?}", a)));
```

Comment:
why does this install a panic hook?

Reply:
This is what failures look like by default:
```
thread 'tests::standard::case_1::cache_file_1_Some__TEST_CACHE_FILE_::cache_dir_1_Some__TEST_CACHE_DIR_::script_paths_file_2_None::warnings_2_None::optimize_2_None::threads_1_Some_4_::no_testonly_2_None::no_breakpoint_1_Some_false_::profile_2_Some_false_' panicked at 'assertion failed: `(left == right)`
  left: `4`,
 right: `8`', scc\tests\args.rs:191:9
```
I think this is readable enough, if you see stack traces it was probably because you set some environment variable that enabled stack printing.

Reply:
if you care about nicely looking errors, I like this crate, you can add it:
https://docs.rs/pretty_assertions/latest/pretty_assertions/